### PR TITLE
Add editable WordArt and 3D shape rotation

### LINKED ISSUE #194: [Contribution FC10] WordArt and 3D rotation
## Status and Destination

Proposed contribution FC10. This issue belongs to **trsdn/mcp-server-ppt** and tracks a potential PR to **sbroenne/mcp-server-powerpoint**. It is not a request to add the feature again here, not maintainer acceptance, and not an archive prerequisite.

## Problem / Use Case

Author editable decorative text and rotated 3D objects where a workflow needs them.

## Component and Target Users

Shared Core implementation with equal MCP Server and CLI behavior. Users: AI assistants and direct CLI automation users.

## Proposed Solution and Scope

Separate if needed; target bevel support already exists and is not a gap.

## Source and Comparison Evidence

Local reference: [src/PptMcp.Core/Commands/Shape/IShapeCommands.cs](https://github.com/trsdn/mcp-server-ppt/blob/d9ec6ab69971b19f94b8b92239755fbbcff0687f/src/PptMcp.Core/Commands/Shape/IShapeCommands.cs). Use this as behavioral reference, not a drop-in port.

Target public-command comparison: `sbroenne/mcp-server-powerpoint` main at `bee71dc`, checked 2026-09-13. No equivalent dedicated operation for this candidate scope was identified in that comparison. Convenience candidates may be composed from existing actions; inspect resources and parameter-level equivalents before finalizing the gap. This is source evidence, not fresh runtime verification.

Local working checklist: docs/CONTRIBUTION-AND-RETIREMENT.md, Feature Contribution Checklist, **FC10**. The checklist is currently a local uncommitted document; this issue includes the scope so it is independently usable.

## Dependencies and Alternatives

No dependency on another contribution is established. Refresh the target and check its open PRs before implementation.

Prefer extending the target's existing owning domain. Consider composing existing actions or declining the contribution if the added API is not worth maintaining. Split this ticket before implementation if grouped operations have materially different contracts.

## Proposed API / Implementation Notes

Finalize action names, parameters, defaults, result shape, and validation against the target's conventions before coding. Use its shared Core interface to generate both entry points. Verify PowerPoint PIA signatures and native behavior; use typed COM ownership with finally cleanup. Do not port session infrastructure or invent unsupported fallback behavior.

## Acceptance Criteria

- [ ] Recheck current target main and open PRs; confirm the gap and record the bounded contract, edge cases, and exclusions.

- [ ] Add a real-PowerPoint regression first, observe failure, then implement and verify the proposed behavior.

- [ ] Cover invalid inputs and meaningful boundary cases; verify unrelated content remains unchanged for mutations.

- [ ] Verify save/reopen for persisted changes, or non-mutation for read-only actions; test output completion/content for export actions.

- [ ] Verify equivalent CLI and MCP parameters, defaults, validation, and results.

- [ ] Pass focused tests with positive executed counts, a zero-warning Release build, and the target-required gates.

- [ ] Synchronize shared guidance, generated references/help, and applicable counts; add the target-required changeset.

- [ ] Link the target PR and report actual checks and limitations. Mark delivered only after merge; record deferral or decline explicitly.

## Tracking

Progress states: proposed -> accepted -> implementing -> PR open -> merged. Deferred and declined are valid alternatives. No implementation or target PR has been started for this ticket.

### PR DESCRIPTION
**Pull request by trsdn**
## Summary
- Adds native editable WordArt and independent per-axis 3D rotation to the Shape API.
- Exposes identical generated MCP and CLI operations.

## Editable WordArt
- New add-text-effect action uses PowerPoint Shapes.AddTextEffect.
- Supports text-effect presets 1 through 50, text/font/font size, bold/italic, and position.
- Produces native editable WordArt, not an image, and persists after save/reopen.

## 3D rotation contract
- New set-3d-rotation and get-3d-rotation actions operate on RotationX, RotationY, RotationZ.
- Partial updates leave omitted axes unchanged.
- X and Y accept inclusive -90 through 90; Z accepts any finite value supported by PowerPoint.
- Requires at least one axis for set.
- Existing 2D Shape.Rotation remains unchanged.

## Generator and integration details
- Adds a ServiceAction override for stable digit-containing action names.
- MCP parameters are snake_case; CLI options are kebab-case.
- Extends ShapeOperationResult with nullable axis values.
- COM references are released in finally blocks.

## Scope
- Bevel behavior is unchanged.
- Full 3D camera, extrusion, depth, lighting, and material APIs are out of scope.

## Testing
- Release solution build: 0 warnings, 0 errors.
- Shape real-PowerPoint integration tests: 71 passed.
- MCP Server tests: 43 passed.
- Release metadata and Agent Skills packaging tests: 77 passed.
- Static checks passed for Success invariants, COM leaks, dynamic casts, interface completeness, documentation counts, and TODO/FIXME/HACK markers.
- Persistence, partial updates, 2D independence, validation boundaries, invalid indices, and generated public schemas are covered.

## Changelog
- Adds a minor changeset for both user-facing capabilities.

Closes trsdn/mcp-server-ppt#194

### CODE REVIEW

**Comment by trsdn** [comment]
Thanks for the review — all findings are addressed in `104c9e9`.

**COM cleanup**
`AddTextEffect`, `Set3DRotation` and `Get3DRotation` no longer reach through property chains. Each now keeps the `PowerPoint.Slides` and `PowerPoint.Shapes` collections in typed locals and releases them in the existing `finally`, alongside the shape and `ThreeDFormat`. Index validation reads `slides.Count` / `shapes.Count` from those locals rather than re-acquiring the collections.

**Preset validation**
`Enum.TryParse` accepted underlying numeric values, so `"0"`, `"1"` or `"49"` bypassed the documented contract. The member-name prefix is now required in addition to the defined-value and `msoTextEffectMixed` checks. Added a theory covering those three inputs, which also asserts no shape is created. Verified red first: all three cases failed before the fix and pass after it.

**CLI contract coverage**
Added two tests to `GeneratedContractTests`:
- `ShapeCli_RoutesWordArtAnd3DRotationActions` — asserts the three actions are in `ValidActions` and that `ServiceRegistry.Shape.RouteCliArgs` returns the expected command strings.
- `Shape3DRotationCli_AcceptsIndividualAxesAndRejectsInapplicableOnes` — routes each axis individually and asserts it reaches the args payload, then asserts `get-3d-rotation` rejects an axis argument and `set-rotation` rejects `rotationZ`, protecting the 2D/3D parameter split.

**Metadata and skill inventory**
- `mcpb/manifest.json` `long_description` now reports 190 operations and mentions the two new capabilities, so the bundle no longer contradicts `mcpb/README.md`.
- `docs/POWERPOINT-NATIVE-FEATURE-AUDIT.md` updated from 187 to 190.
- `skills/powerpoint-mcp/SKILL.md` now lists `add-text-effect`, `set-3d-rotation` and `get-3d-rotation` in the top-level inventory, matching the CLI skill and the packaged references.

**Validation** (full `scripts/pre-commit.ps1`, exit 0)
- Release build: 0 warnings, 0 errors
- Shape real-PowerPoint tests: 74 passed
- MCP Server tests: 45 passed
- Release metadata and Agent Skills packaging tests: 77 passed
- Success-flag, COM-leak, dynamic-cast, interface-completeness and documentation-count gates all passed

**Comment by trsdn** [comment]
Second round addressed in `4b9354a`.

**Non-finite numeric inputs — fixed**
Correct catch. `fontSize <= 0` is false for both `NaN` and `+Infinity`, so those reached `Shapes.AddTextEffect`. `left` and `top` had the same gap, so all three are now required to be finite before entering COM, matching the checks already used by the 3D rotation axes. Added two theories (`NaN`, `+Infinity`, `-Infinity` for the size; `NaN`/`+Infinity` for the position), each also asserting no shape was created. Verified red first — 3/3 failed before the fix, 11/11 pass after.

**`PowerPoint.ThreeDFormat` — not a compile issue, declining this one**
This declaration does compile, and the type is the correct one. `Microsoft.Office.Interop.PowerPoint.ThreeDFormat` exists in the referenced PIA and is exactly what `Shape.ThreeD` returns. Verified by reflecting over the referenced assembly `microsoft.office.interop.powerpoint/15.0.4420.1018/lib/net20/Microsoft.Office.Interop.PowerPoint.dll`:

```
ThreeDFormat types in PowerPoint PIA: Microsoft.Office.Interop.PowerPoint.ThreeDFormat
Microsoft.Office.Interop.PowerPoint.Shape.ThreeD -> Microsoft.Office.Interop.PowerPoint.ThreeDFormat
```

`Microsoft.Office.Core.ThreeDFormat` is the corresponding type on `Microsoft.Office.Core.Shape` in the MSO Core library, which is a different interface hierarchy. Switching to the `Office` alias here would actually break the build. The Release build passes with 0 warnings and 0 errors, which independently confirms the current declaration binds correctly.

**The three inline comments carried over from the first review** (`ShapeCommands.cs` Shapes cleanup, `mcpb` 187 operations, MCP skill inventory) were all resolved in `104c9e9`; they appear to have been re-anchored onto the new commit rather than re-raised. Current state:
- `Set3DRotation` / `Get3DRotation` hold `slides` and `shapes` in typed locals and release both in `finally`.
- `mcpb/manifest.json` reports 190 operations.
- `skills/powerpoint-mcp/SKILL.md` lists all three new actions.

**Validation** (full `scripts/pre-commit.ps1`, exit 0)
- Release build: 0 warnings, 0 errors
- Shape real-PowerPoint tests: 79 passed
- MCP Server tests: 45 passed
- Release metadata and Agent Skills packaging tests: 77 passed
- Success-flag, COM-leak, dynamic-cast, interface-completeness and documentation-count gates all passed

File: src/PowerPointMcp.Core/Shape/ShapeCommands.cs [hunk 1]
```diff
@@ -649,6 +737,131 @@ public ShapeOperationResult GetRotation(IPresentationBatch batch, int slideIndex
...
+            return new ShapeOperationResult { Success = false, ErrorMessage = "3D rotation values must be finite numbers." };
+        }
+
+        return batch.Execute((ctx, ct) =>
+        {
+            PowerPoint.Slides? slides = null;
+            PowerPoint.Slide? slide = null;
+            PowerPoint.Shapes? shapes = null;
+            PowerPoint.Shape? shape = null;
+            PowerPoint.ThreeDFormat? threeD = null;
```
[the remark this answers is not in the document]
Comment (trsdn) [reply src/PowerPointMcp.Core/Shape/ShapeCommands.cs:792]: This one is factually incorrect, so I am not making this change — the type is already correct and the project compiles.

`Microsoft.Office.Interop.PowerPoint.ThreeDFormat` **does** exist, and it is exactly what `Shape.ThreeD` returns. Reflected over the referenced assembly (`microsoft.office.interop.powerpoint/15.0.4420.1018/lib/net20/Microsoft.Office.Interop.PowerPoint.dll`):

```
ThreeDFormat types in PowerPoint PIA: Microsoft.Office.Interop.PowerPoint.ThreeDFormat
Microsoft.Office.Interop.PowerPoint.Shape.ThreeD -> Microsoft.Office.Interop.PowerPoint.ThreeDFormat
```

`Microsoft.Office.Core.ThreeDFormat` is the type on `Microsoft.Office.Core.Shape`, a different interface hierarchy reached through the `OfficeInterop` alias. Switching these declarations to `Office.ThreeDFormat` would introduce the compile error this comment predicts.

Independent confirmation: the Release build of the full solution completes with 0 warnings and 0 errors on every commit in this PR, and the `CI Gate` check is green. The bevel code uses `dynamic` for historical reasons, not because a typed PowerPoint `ThreeDFormat` is unavailable.

**Comment by trsdn** [comment]
Third round addressed in `f3969f1`.

**`rotation_z` wording — fixed**
Correct, and the wrong wording was mine. `ThreeDFormat.RotationZ` rotates the shape around its Z axis; there is no effects camera involved. Updated the canonical `skills/shared/slides-and-shapes.md` and both generated reference copies to: *"Z rotates the shape around its Z axis and is tracked separately from the shape's 2D `rotation`."* The `IShapeCommands` XML docs never contained the camera claim, so interface and guides now agree.

**WordArt formatting coverage — fixed**
`AddTextEffect_CreatesEditableWordArt_AndPersistsAfterSave` now passes `bold: true, italic: false` and, after save and reopen, asserts the persisted `TextEffect.FontName`, `FontSize`, `FontBold` and `FontItalic`. The flags are deliberately asymmetric, so a swapped bold/italic argument fails the test rather than passing silently.

**`PowerPoint.ThreeDFormat` — still declining, with evidence**
Answered in the review thread as well. `Microsoft.Office.Interop.PowerPoint.ThreeDFormat` exists in the referenced PIA and is the declared return type of `Shape.ThreeD`:

```
ThreeDFormat types in PowerPoint PIA: Microsoft.Office.Interop.PowerPoint.ThreeDFormat
Microsoft.Office.Interop.PowerPoint.Shape.ThreeD -> Microsoft.Office.Interop.PowerPoint.ThreeDFormat
```

Using the `Office` alias here would bind `Microsoft.Office.Core.ThreeDFormat`, which belongs to `Microsoft.Office.Core.Shape` — that is the change that would break the build. The Release build is 0 warnings / 0 errors and `CI Gate` is green on every commit in this PR.

**Two carried-over inline comments are stale.** `mcpb/manifest.json` has reported 190 operations since `104c9e9`, and `skills/powerpoint-mcp/SKILL.md` lists `add-text-effect`, `set-3d-rotation` and `get-3d-rotation` on lines 103-104. Both were verified in the working tree at the current head.

**Validation** (full `scripts/pre-commit.ps1`, exit 0)
- Release build: 0 warnings, 0 errors
- Shape real-PowerPoint tests: 79 passed
- MCP Server tests: 45 passed
- Release metadata and Agent Skills packaging tests: 77 passed
- Success-flag, COM-leak, dynamic-cast, interface-completeness and documentation-count gates all passed

**Comment by trsdn** [comment]
Fourth round addressed in `eb54ff5`. All four points were valid.

**At-least-one-axis contract — documented**
`IShapeCommands.Set3DRotation` now states *"At least one axis must be supplied; omitted axes remain unchanged."*, so the generated MCP and CLI descriptions carry the same rule the implementation enforces. The canonical `skills/shared/slides-and-shapes.md` and both mirrored reference copies were updated to match.

**TextEffect RCW released**
The persistence assertion now walks `Slides → Slide → Shapes → Shape → TextEffect` through typed locals and releases all five in a `finally` with `ComUtilities.Release`, matching the pattern already used in `ChartCommandsTests.SetRawChartVariant`.

**Boundary and non-finite coverage added**
- `Set3DRotation_WithInclusiveRangeEndpoints_Succeeds` — `(-90, 90)` and `(90, -90)`; asserts success and that both values are returned unchanged, confirming the endpoints really are inclusive.
- `Set3DRotation_WithNonFiniteAxis_ReturnsFailure` — `NaN` and an infinity on each of X, Y and Z individually, including the otherwise unrestricted Z.

The existing out-of-range theory is unchanged, so the range check and the finite check are now both pinned.

**Validation** (full `scripts/pre-commit.ps1`, exit 0)
- Release build: 0 warnings, 0 errors
- Shape real-PowerPoint tests: 87 passed (was 79; 8 new cases)
- MCP Server tests: 45 passed
- Release metadata and Agent Skills packaging tests: 77 passed
- Success-flag, COM-leak, dynamic-cast, interface-completeness and documentation-count gates all passed

The only open item I am not acting on is the `PowerPoint.ThreeDFormat` namespace comment, answered with reflection output from the referenced PIA in the thread above.

**Comment by trsdn** [comment]
Fifth round addressed in `ea696de`. The MCP server README shape row now reads *"shape creation, editable WordArt, styling, per-axis 3D rotation, grouping, hyperlinks, placeholders, string tags, ..."*, so the 50-operation claim and the listed capabilities agree. Pre-commit passed (exit 0); documentation count validation confirms 16 tools / 190 operations.

**Comment by trsdn** [comment]
Rebased onto current `main` via a merge commit (`2ea1186`) — the branch was conflicting after #87 landed.

**What the merge had to reconcile**
#87 touched the same Shape test files, action lists and operation counts as this branch.

- **Tests:** both suites are kept. `GeneratedContractTests` now has the copy-formatting wiring test alongside the two WordArt/3D routing tests, and the two MCP schema tests (`ShapeSchema_ExposesCopyFormattingActionAndItsShapeIndexes` and `ShapeSchema_ExposesWordArtAnd3DRotationActions`) each extract the action enum independently instead of sharing one block.
- **Counts:** upstream 188 operations / shape 48, plus the three actions this PR adds, gives **191 operations / shape 51**. Updated consistently across the README, feature reference, installation and MCP docs, the bundle README and manifest, the audit, the workflow guides and the MCP server README.
- **Action lists:** the generated `skills/powerpoint-cli/SKILL.md` and `references/cli-commands.md` were regenerated by the build and now carry `copy-formatting` together with `add-text-effect`, `set-3d-rotation` and `get-3d-rotation` in generator order. The hand-maintained `gh-pages/docs/features.md` and `skills/powerpoint-mcp/SKILL.md` were updated to match.
- Both changesets (`shape-copy-formatting.md` and `bright-shapes-rotate.md`) are present.

**Validation after the merge** (full `scripts/pre-commit.ps1`, exit 0)
- Release build: 0 warnings, 0 errors
- Shape real-PowerPoint tests: 95 passed
- MCP Server tests: 47 passed
- Release metadata and Agent Skills packaging tests: 77 passed
- Documentation count validation: 16 tools / 191 operations
- Success-flag, COM-leak, dynamic-cast and interface-completeness gates all passed

The previous round's findings remain addressed; nothing in the merge reverted them.

**Comment by trsdn** [comment]
Both nits fixed in `4bd8675`. The CLI reference one was a real gap — thanks for catching it.

**`skills/powerpoint-cli/references/cli-commands.md` — regenerated**
You were right: the checked-in file contained `copy-formatting` but none of the new shape options. It had been generated before the new commands were compiled, so the merge carried a stale copy. Regenerated from the Release `pptcli` via `scripts/Build-AgentSkills.ps1 -PopulateReferences`. It now documents `add-text-effect` (14 occurrences including the `--preset-effect`, `--font-name`, `--font-size`, `--bold` and `--italic` flags), `set-3d-rotation` and `get-3d-rotation` with `--rotation-x`/`--rotation-y`/`--rotation-z`, and still documents `copy-formatting`.

**`gh-pages/docs/index.md` — count made consistent**
The transcript at line 75 spelled out "One hundred eighty-six operations", which predates both this PR and #87. Changed to "One hundred ninety-one operations" so the page agrees with the CTA and the feature reference.

**Validation:** pre-commit passed (exit 0); documentation count validation reports 16 tools / 191 operations. The previous full run on the merge commit covered Release build 0/0, Shape 95, MCP Server 47 and packaging 77.
